### DISCUSSION TITLE
Creating tiled images via a .bat file

### DISCUSSION THREAD

**Discussion by Trailboy** [ImageMagick/ImageMagick#1914] [General]
Hi. 

Hope you can all help me!

Version: ImageMagick 7.0.8-33 Q16 x86 2019-03-10 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC Modules OpenMP

I have been using imagemagick to create tiled prints.  

I have used it via a windows .bat file to quickly create an image that is 6x4 inch equivalent in pixels that have on them either 4 3x2 inch prints or 16 1.5x1 inch prints.  

They need to be .bat files so that when I just drop a .jpg image on to it creates it for me as is has to be created quickly when I need it (customer waiting).   

Keyrings-6x4-16on

`magick convert ( %1 -rotate "-90>" -resize 300 -gravity NorthWest -extent 1200x1800 ) ( %1 -rotate "-90>" -resize 300 -write mpr:small ) -gravity NorthWest -geometry +300+0 -composite mpr:small -gravity NorthWest -geometry +600+0 -composite mpr:small -gravity NorthWest -geometry +900+0    -composite mpr:small -geometry +0+450 -composite mpr:small -geometry +300+450 -composite mpr:small -geometry +600+450 -composite mpr:small -geometry +900+450 -composite mpr:small -geometry +0+900 -composite mpr:small -geometry +300+900  -composite mpr:small -geometry +600+900 -composite mpr:small -geometry +900+900 -composite mpr:small -geometry +0+1350 -composite mpr:small -geometry +300+1350 -composite mpr:small -geometry +600+1350 -composite mpr:small -geometry +900+1350 -composite   "%C:\Users\Kris\Desktop\ImageMagick\%~n1_Keyrings-6x4-16on.jpg"`

Magnets-6x4-4on

`magick convert ( %1 -rotate "-90>" -resize 600 -gravity NorthWest -extent 1200x1800 ) ( %1 -rotate "-90>" -resize 600 -write mpr:small ) -gravity NorthWest -geometry +600+0 -composite mpr:small -geometry +600+900 -composite mpr:small -geometry +0+900 -composite "%C:\Users\Kris\Desktop\ImageMagick\%~n1_Magnets-6x4-4on.jpg"`

THis has worked out for me well so far. My questions are:

1. Am I doing this right or is there an easier methodology?  I think the original was created some time ago and not by me, but has been amended.

2. Can I create a single 1200x1800 pixel image that has 2 900x600 pixel (i.e. 3x2 inch) images on one half and 8 450x300 (1.5x1 inch) images on the other half?

3.  If so, is it possible to create even more complex patterns, perhaps a 2700x1800 (i.e. 9x6 inch) image that has perhaps 1 1800x1200 (6x4 inch), 2 900x600 images and the rest with 450x300 perhaps?  

4. Is is possible to leave a very slight white border around the outside of the completed image to compensate for the fact that my printer crops slightly from the edge?

5.  Is it possible to code it such that the image is created in whatever folder the .bat script resides in?  I have to use a particular folder as above in the code, but if it could automatically default to the same folder it would help me enormously.  

Sorry for the long winded questions!  I'm learning rapidly!  

Thanks for any advice.

Kris

**Comment by snibgo** [comment]
1. To make each command easier to understand, I suggest you split it into lines. Put a caret ^ at the end of each line except the last. For example:
[code]magick ^
  ( %1 -rotate "-90>" -resize 300 -gravity NorthWest -extent 1200x1800 ) ^
  ( %1 -rotate "-90>" -resize 300 -write mpr:small ) ^
  -gravity NorthWest -geometry +300+0 -composite ^
  mpr:small -gravity NorthWest -geometry +600+0 -composite ^
  mpr:small -gravity NorthWest -geometry +900+0 -composite ^
  mpr:small -geometry +0+450 -composite ^
  mpr:small -geometry +300+450 -composite ^
  mpr:small -geometry +600+450 -composite ^
  mpr:small -geometry +900+450 -composite ^
  mpr:small -geometry +0+900 -composite ^
  mpr:small -geometry +300+900 -composite ^
  mpr:small -geometry +600+900 -composite ^
  mpr:small -geometry +900+900 -composite ^
  mpr:small -geometry +0+1350 -composite ^
  mpr:small -geometry +300+1350 -composite ^
  mpr:small -geometry +600+1350 -composite ^
  mpr:small -geometry +900+1350 -composite ^
  "%~n1_Keyrings-6x4-16on.jpg"
[/code]

Each command is more complex than it needs to be. "-duplicate" and "+append" and "-append" would be useful to simplify the command. In addition, each command reads the input, rotates and resizes it twice. The command can be optimised so it does this only once.

For v7, I suggest you use "magick", not "magick convert".

2. Yes, you can. "-resize" images, and append as required.

3. Yes, as (2).

4. "-bordercolor white -gravity center -border 20" will add a 20-pixel white border.

5. "%~dp0" is the drive and directory where the script file resides in. Use that instead of "C:\Users\Kris\Desktop\ImageMagick\".

5a. You have a leading "%" in the output file, eg "%C:\Users...". What is that for?

**Reply by Trailboy** [reply]
Thank you!

**Comment by Trailboy** [comment]
Thank you snibgo

So, I'm making progress with this (more than I have in the last two weeks anyway).

Keyrings script:

```
magick ^
	( %1 -rotate "-90>" -resize 300 -gravity NorthWest -extent 1200x1800 ) ^
	( %1 -rotate "-90>" -resize 300 -write mpr:small ) ^
	-gravity NorthWest -geometry +300+0 -composite ^
	mpr:small -gravity NorthWest -geometry +600+0 -composite ^
	mpr:small -gravity NorthWest -geometry +900+0 -composite ^
	mpr:small -geometry +0+450 -composite ^
	mpr:small -geometry +300+450 -composite ^
	mpr:small -geometry +600+450 -composite ^
	mpr:small -geometry +900+450 -composite ^
	mpr:small -geometry +0+900 -composite ^
	mpr:small -geometry +300+900  -composite ^
	mpr:small -geometry +600+900 -composite ^
	mpr:small -geometry +900+900 -composite ^
	mpr:small -geometry +0+1350 -composite ^
	mpr:small -geometry +300+1350 -composite ^
	mpr:small -geometry +600+1350 -composite ^
	mpr:small -geometry +900+1350 -composite ^
	-bordercolor white -gravity center -border 20 ^
	"%~dp0\%~n1_Keyrings-6x4-16on.jpg"
```

Magnets Script:

```
magick ^
	( %1 -rotate "-90>" -resize 600 -gravity NorthWest -extent 1200x1800 ) ^
	( %1 -rotate "-90>" -resize 600 -write mpr:small ) ^
	-gravity NorthWest -geometry +600+0 -composite ^
	mpr:small -geometry +600+900 -composite ^
	mpr:small -geometry +0+900 -composite ^
	-bordercolor white -gravity center -border 20 ^
	"%~dp0\%~n1_Magnets-6x4-4on.jpg"
```
Will have to do a lot more research on the duplicate, resize and append commands.  

Re the leading "%" - not sure how this was added, but is obviously superfluous.

**Comment by Trailboy** [comment]
Hi again.  

Sorry to bother you yet again.  

Experimenting, and so far have come up with the following amendments (with help from a little forum research):

Keyrings Script:

```
magick %1 ^
	-set filename:f "%%[t]_3x2" ^
	-gravity center ^
	-extent "%%[fx:w<h&&w/2>h/3?h/3*2:w]x%%[h]" ^
	-extent "%%[fx:w>h&&w/3>h/2?h/2*3:w]x%%[h]" ^
	-extent "%%[w]x%%[fx:w<h&&w/2<h/3?w/2*3:h]" ^
	-extent "%%[w]x%%[fx:w>=h&&w/3<h/2?w/3*2:h]" ^
	-resize 450x450 ^
	-quality 68 ^
	-duplicate 3 +append ^
	-duplicate 3 -append ^
	-bordercolor white -gravity center -border 10 ^
	"%~dp0\%~n1_Keyrings-Testing.jpg"
```

Magnets Script:

```
magick %1 ^
	-set filename:f "%%[t]_3x2" ^
	-gravity center ^
	-extent "%%[fx:w<h&&w/2>h/3?h/3*2:w]x%%[h]" ^
	-extent "%%[fx:w>h&&w/3>h/2?h/2*3:w]x%%[h]" ^
	-extent "%%[w]x%%[fx:w<h&&w/2<h/3?w/2*3:h]" ^
	-extent "%%[w]x%%[fx:w>=h&&w/3<h/2?w/3*2:h]" ^
	-resize 900x900 ^
	-quality 68 ^
	-duplicate 1 +append ^
	-duplicate 1 -append ^
	-bordercolor white -gravity center -border 10 ^
	"%~dp0\%~n1_Magnets-Testing.jpg"
```

which both appear to do what I want. Great and gets me moving along for now.

A few questions left:

1. I think this a better syntax, but am I going down the wrong path?  Especially relating to the next question...

2. How might I go about making more complex patterns which include a larger initial canvas with multiple different sizes of images on?  I suspect this is above my limited understanding for now.  

3.   The code '"%~dp0\%~n1_Magnets-Testing.jpg"' gives me the output as the same directory as the .bat script.  Is there a similar shortcut to create the output in the same directory as the source image?  

Thanks for helping me.

**Comment by snibgo** [comment]
1. You have "-set filename:f "%%[t]_3x2" ^" but then never use "filename:f", so you can remove that line.

2. Think of the "multiple different sizes" as a number of images, each made from the same input. There are different ways of doing this. For example, with clone and parentheses, we make a copy resized to 600 and duplicate that, then make another resized to 300, and finally append them all sideways:

magick ^
  %1 ^
  ( -clone 0 -resize 600 -duplicate 2 ) ^
  ( -clone 0 -resize 300 ) ^
  +append ^
  out.png

3. The script name is %0, and the first argument is %1. So use "%~dp1" instead of "%~dp0".
